### DISCUSSION TITLE
Install chrome on a Forge provisioned Ubuntu 20.04 server

### DISCUSSION THREAD

**Discussion by Loots-it** [spatie/browsershot#484] [Q&A]
Hi, 

in the documentation you provide a way to install chrome on a forge provisioned Ubuntu 16.04 version. Will it also work on ubuntu 20.4?

I'm also wondering if this packages launches a new puppeteer instance on every pdf generation or reuses old instances? Because timing and memory usage is important to me. And if the same instance is reused I probably need a supervisor in case the instance crashes.

I have one last question regarding local development, what's the best way to install chrome for laravel sail?

If I find any of the answers myself I will comment on this discussion.

**Comment by Loots-it** [comment]
I tried to install chrome on a forge provisioned Ubuntu 20.4 server with the instructions from the documentation. It seems to work just fine.

I think the package reuses the old puppeteer instance if an instance is running, but I'm not 100% sure.

I wasn't able to use this package with laravel sail (yet).

**Comment by garbinmarcelo** [comment]
@Loots-it Did you manage to use it with Sail or gave up?

**Reply by Loots-it** [reply]
I gave up, I do have a staging server to test an exact replica of my production environment. I also use some tests that I run locally (not via docker/sail). It works, but it would be nice if I could use it with sail. So did you find anything to make it work by any chance?

**Reply by Loots-it** [reply]
Maybe this helps: https://stackoverflow.com/questions/66708854/laravel-sail-install-puppeteer-chromium

I haven’t tested it though, if you find the time to try it I would love to hear the results 😉

**Reply by garbinmarcelo** [reply]
> Maybe this helps: https://stackoverflow.com/questions/66708854/laravel-sail-install-puppeteer-chromium
> 
> I haven’t tested it though, if you find the time to try it I would love to hear the results 😉

I just did this, and it didn't work. I'm still looking to use it with Sail, I wasted a lot of time with this error, I was trying to run an application with Sail and it always gave an error, until I ran it directly on Ubuntu from WSL and it worked, I even got a bit annoyed with it hehehe  because I lost a long time... That's why I'm looking to find a solution with Sail.

If I have news, I'll answer you here.

**Reply by ap1969** [reply]
Hi, for sail, try the following:

```
Browsershot::url($url)
            ->setRemoteInstance('172.22.0.100', '9222')
            -> .....
```

And modify the sail docker-compose as follows (I've highlighted the key points):

```
# For more information: https://laravel.com/docs/sail
version: "3"
services:
    ## ADD THE CHROME IMAGE AS A SERVICE
    chromium:
        image: zenika/alpine-chrome
        command:
            [
                chromium-browser,
                "--headless",
                "--disable-gpu",
                "--remote-debugging-address=0.0.0.0",
                "--remote-debugging-port=9222",
            ]
        cap_add:
            - SYS_ADMIN
        ports:
            - "9222:9222"
        networks:
            sail:
                ipv4_address: 172.22.0.100
 ## MODIFY THE NETWORK
networks:
    sail:
        ipam:
            driver: default
            config:
                - subnet: "172.22.0.1/24"
volumes:
    sail-mysql:
        driver: local
    sail-redis:
        driver: local
    sail-meilisearch:
        driver: local
    sail-minio:
        driver: local

```

I've got this working on my windows PC fine. Just need to work out the best way to enable/disable the setRemoteInstance based on the environment.

The only problem I've found with this is that I want to use the Browsershot::html() function, and I've not yet got that working with the remote instance.


Hope that helps,
Andy

**Reply by ap1969** [reply]
Further details here: https://github.com/spatie/browsershot/discussions/711
